Re-land Expose height and width factor in AnimatedAlign

Nash0x7E2 opened the pull request:
Reland #60836 following #61109. 

cc: @Piinks 

## Description
`Align` allows developers to customize the width and height factor, however `AnimatedAlign` does not expose these properties. This PR exposes `widthFactor` and `heightFactor` for AnimatedAlign and adds the necessary Tween visitors and documentation for both properties.

## Tests
No test added. I looked at the existing implementation of `Align` and saw that there was no test covering the height and width factors so none was added to the Animated widget. 

## Checklist

- [ x ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ x ] I signed the [CLA].
- [ x ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ x ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ x ] I updated/added relevant documentation (doc comments with `///`).
- [ x ] All existing and new tests are passing.
- [ x ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ x ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [ x ] No, no existing tests failed, so this is *not* a breaking change.

Nash0x7E2 committed:
Expose height and width factor

- Expose the height and width factor properties in AnimatedAlign

Nash0x7E2 committed:
Add type information to tween

Nash0x7E2 committed:
add review suggestions

Nash0x7E2 committed:
add test for animated align

Nash0x7E2 committed:
add height/width factor test to align

Nash0x7E2 committed:
minor doc update

Nash0x7E2 committed:
Fix indentation :)

Co-authored-by: Kate Lovett <katelovett@google.com>

Piinks commented on packages/flutter/test/widgets/align_test.dart:
```diff
@@ -112,4 +112,70 @@ void main() {
+          mainAxisAlignment: MainAxisAlignment.center,
+          crossAxisAlignment: CrossAxisAlignment.start,
+          children: <Widget>[
+            Align(
```
Why are there two instance of Align here and in the animated tests?

Piinks commented on packages/flutter/test/widgets/align_test.dart:
```diff
@@ -112,4 +112,70 @@ void main() {
+        ),
+      ),
+    );
+    final RenderBox box = inner.currentContext.findRenderObject() as RenderBox;
```
For all, these lookups could be:
```dart
final RenderBox textBox1 = tester.renderObject<RenderBox>(find.byKey(key1));
```
Or if there were only one instance of the class:
```dart
final RenderBox box = tester.renderObject<RenderBox>(find.byType(AnimatedAlign));
```

Piinks reviewed (commented):
Hey @Nash0x7E2 thanks for re-opening this! The test failures on the other side were a bit tricky to understand, but I think Mehmet was right.
`Align` can have null values for the height and width factors, and does so by default, so we should try to match that here. It looks like in the underlying render object, `RenderPositionedBox`, the width and height factors indicate if we are shrink wrapping, so even if the factor is 1.0, it can have a very different result.
Here's the render object:
https://github.com/flutter/flutter/blob/c29009b3e89cc4343d318ef18d6cbe4e6577cef7/packages/flutter/lib/src/rendering/shifted_box.dart#L392

Once updated, I can initiate the tests before landing to make sure it isn't breaking again. :)

Nash0x7E2 committed:
height and width factor should be null

Signed-off-by: Neevash Ramdial <mail@neevash.dev>

Nash0x7E2 committed:
apply tween if values are not null

Signed-off-by: Neevash Ramdial <mail@neevash.dev>

Nash0x7E2 committed:
updates AnimatedAlign test with suggestions

Signed-off-by: Neevash Ramdial <mail@neevash.dev>

Nash0x7E2 committed:
remove extra widgets (Align test)

Signed-off-by: Neevash Ramdial <mail@neevash.dev>

Nash0x7E2 commented:
Hi @Piinks Thank you for the detailed explanation! I did not know about the underlying shrink wrapping, definitely learned something new :) I've implemented your suggestion and updated the appropriate test.

AyushBherwani1998 committed:
[Api docs] Updates ListTile.subtitle docs to address how TextStyle is decided (#61071)

# Conflicts:
#	packages/flutter/lib/src/material/list_tile.dart

Piinks reviewed (approved):
LGTM! I'm going to run this through the extra tests and report back. 👍 
I'm applying WIP so it won't be merged by accident while I check.

Piinks commented:
Re: the analyzer failures, they were from a break on master. Updating your branch should resolve them. :)

Nash0x7E2 commented:
Ok cool, fingers crossed! Thanks @Piinks :)

Piinks commented:
Tests passed! 🎉 
Once you update your branch we can re-land this! Thanks for your continued work on it!

Nash0x7E2 commented:
Ok cool, updating my branch now

Nash0x7E2 commented:
@Piinks cherry-picked the latest commits from master

Piinks commented:
Ah I see! Github doesn't handle some rebasing methods very well. Right now it thinks there are multiple authors so the cla is failing in addition to the original analyzer issue. 

I usually update my branch with
```git
git fetch upstream
git pull upstream master
git push origin $branch
```

I don't often have an issue with that, but there are other steps we can take to clear this up if that does not work. 👍

Nash0x7E2 committed:
Merge branch 'master' of https://github.com/flutter/flutter into nash/animated-align
# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

Nash0x7E2 commented:
I see, tried it. Hopefully this works

Piinks commented:
Awesome! Looks like that cherry picked commit stuck around, but everything else is resolved. I haven't seen that happen before! 🤔  
I think I can override the cla in this case. Let me just check that this won't disturb the git history.

Nash0x7E2 commented:
Awesome, thank you :)
